### PR TITLE
feedback: send default metrics

### LINKED ISSUE #8192: Send metrics in feedback daemon
[PM link](https://www.notion.so/tarantool/metrics-min-avg-max-percentile-50-95-99-56433be43d5f4e55a10c81d70fd0ad4e)

Introduce new cfg options (defaults):
`feedback_send_metrics`(true) - collect and send metrics only if the flag is set,
`feedback_metrics_collect_interval`(60) - period of metrics collection, in seconds,
`feedback_metrics_limit`(1048576) - memory limit for metrics, in bytes.

If feedback is enabled and feedback_send_metrics is true, a background fiber must be started that collects metrics every feedback_metrics_collect_interval seconds. It must call `require('metrics').collect()` but it must do nothing if metrics module is not found. All the metrics must be appended to a huge array, but the approximate size exceeds feedback_metrics_limit bytes, some earlier collected metrics must be dropped.

If feedback_send_metrics is true, all the collected metrics must be send with other feedback to feedback server. At this moment all collected metrics must be dropped.

I think that the most complex part of this issue is to test it: we should somehow create a feedback server and check that all the metrics are successfully delivered to it.

### PR DESCRIPTION
**Pull request by drewdzzz**
The patch extends feedback_daemon functionality - now it collects default metrics if module metrics of required version (>= 0.16.0) is installed. The metrics are sent as a part of feedback. Feedback version is bumped.

Closes #8192

Manually checked that it works with actual `metrics` module.

### CODE REVIEW

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [inline src/box/lua/feedback_daemon.lua:406]: Where did you get the numbers for approximate type sizes from?

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [inline src/box/lua/feedback_daemon.lua:457]: Not sure this is correct. Since the intervals are different, technically, I guess there should be two guard loops.

File: test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
Comment (CuriousGeorgiy) [inline test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua:61]: How can there be `metrics`, if you explicitly unload it in `g.before_each`?

File: test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
Comment (CuriousGeorgiy) [inline test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua:52]: Why don't you set a big enough `feedback_metrics_collect_interval` to test both cases?

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [inline src/box/lua/feedback_daemon.lua:10]: I would add a separate patch adding some prefix, e.g., `feedback` to `PREFIX`, `fiber`, etc., because now it becomes unclear what they refer to.

**Review by CuriousGeorgiy — CHANGES_REQUESTED** [review]
Is there any integration test with `tarantool/feedback` or `tarantool/metrics` planned to test this patch?

File: src/box/lua/load_cfg.lua
Comment (CuriousGeorgiy) [inline src/box/lua/load_cfg.lua:185]: But the PRD [says](https://www.notion.so/tarantool/metrics-min-avg-max-percentile-50-95-99-56433be43d5f4e55a10c81d70fd0ad4e#fdd6a1690d1a4710afe50dff6de9d87e) it should be 3600?

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [reply src/box/lua/feedback_daemon.lua:419]: Your implementation also contradicts PRD.

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:406]: From the most popular website for Lua programmers - [Gamepedia](https://wowpedia.fandom.com/wiki/Lua_object_memory_sizes)!
But seriously, it is the only place where I can get approximate sizes of objects. Another place is Lua source code, but it is much harder to extract such information from there.

**Comment by drewdzzz** [comment]
@CuriousGeorgiy Integration test is not planned

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:419]: > What if metrics_size is always larger than the size of one metric? Shouldn't we keep at least one metric?

As for me - no. If we never fit the limit, we should not store metrics.

File: test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
Comment (drewdzzz) [reply test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua:61]: `require` does more than reading from `package.loaded` - search for the module file in current and user dirs, for example. So, unloading the module does not help here.

File: test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
Comment (drewdzzz) [reply test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua:52]: Feedback daemon is cleared on reconfiguring, so I will need to wait for a long time before the first collection.1

File: src/box/lua/load_cfg.lua
Comment (drewdzzz) [reply src/box/lua/load_cfg.lua:185]: > `box.cfg.feedback_metrics_collect_interval = 3600`
Позволяет задавать интервал сбора метрик. 
По умолчанию, раз в минуту.

3600 is not a minute.

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [reply src/box/lua/feedback_daemon.lua:419]: @artur-barsegyan RFC

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:10]: Could you elaborate, please? Why do I need a separate patch? It is just name of the fiber for logs.

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:406]: Agree

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:457]: I'm not even sure that we need this guard loop.

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [reply src/box/lua/feedback_daemon.lua:10]: After this patch we will the daemon will have fields like `fiber` and `metrics_collect_fiber` — I suggest you add the prefix `feedback` to the latter ones for clarity.

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:457]: I don't see its value in reconfiguration. I assume that it is needed to restart the fiber if it fails with an error. However, I don't see any calls that can cause an error.

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:406]: Discussed with @igormunkin - generally it is OK for approximate size count. But if gc64 is enabled, trivial types consumes no memory - they are stored right in table slot. Added this logic.

<details>
<summary>diff</summary>

```diff

diff --git a/src/box/lua/feedback_daemon.lua b/src/box/lua/feedback_daemon.lua
index 4c8e292c9..58ac8c64b 100644
--- a/src/box/lua/feedback_daemon.lua
+++ b/src/box/lua/feedback_daemon.lua
@@ -5,6 +5,7 @@ local json  = require('json')
 local fiber = require('fiber')
 local http  = require('http.client')
 local fio = require('fio')
+local ffi = require('ffi')
 
 local PREFIX = "feedback_daemon"
 local METRICS_PREFIX = "metrics_collector"
@@ -386,11 +387,12 @@ local function collect_default_metrics()
     end
 end
 
+-- Trivial objects are stored right in table slot or stack with gc64
+local trivial_obj_size = ffi.abi('gc64') and 0 or 8
+
 -- Calculate approximate amount of occupied memory
 local function obj_approx_size(obj)
-    if type(obj) == nil then
-        return 0
-    elseif type(obj) == 'table' then
+    if type(obj) == 'table' then
         local size = 40
         for k, v in pairs(obj) do
             size = size + 16
@@ -400,10 +402,8 @@ local function obj_approx_size(obj)
         return size
     elseif type(obj) == 'string' then
         return 17 + #obj
-    elseif type(obj) == 'boolean' then
-        return 1
-    else -- Numbers and cdata
-        return 8
+    else -- Number, boolean and nil
+        return trivial_obj_size
     end
 end
 
diff --git a/test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua b/test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
index e0fa83612..3dc08e8df 100644
--- a/test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
+++ b/test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
@@ -120,17 +120,16 @@ g.test_memory_limit = function()
     g.server:exec(function()
         local t = require('luatest')
         local fiber = require('fiber')
-        box.cfg{feedback_metrics_limit = 8 * 5}
+        box.cfg{feedback_metrics_limit = 20 * 5}
         local m = {}
         m.collect = function(_)
-            return 0
+            return "abc"
         end
         m._VERSION = "0.16.0"
         package.loaded["metrics"] = m

```
</details>

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:11]: I've added `send` prefix to some fields in a separate commit.

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [reply src/box/lua/feedback_daemon.lua]: Thanks for the research!

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:419]: Discussed with @artur-barsegyan - we should always fit the limit.

**Comment by drewdzzz** [comment]
@CuriousGeorgiy 
I decided to change preemption algo - now I keep old ones. Turned out that it does not matter which ones to keep, but keeping the old ones is simplier, does not require additional array of sizes and we does not need to move elems in huge array of metrics.

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [inline src/box/lua/feedback_daemon.lua:12]: Nit, feel free to resolve and ignore: I would put this commit first.

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [inline src/box/lua/feedback_daemon.lua:405]: Nit, feel free to resolve and ignore: I would open a ticket for refactoring this part into `src/lua/utils.h`, as this may be useful in other cases.

File: test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
Comment (CuriousGeorgiy) [reply test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua:53]: I don't get your point, could you please elaborate? AFAIU, metrics are collected every `feedback_metrics_collect_interval`, and you explicitly retrieve them by calling `daemon.generate_feedback` — so why will you have to wait long?

**Review by CuriousGeorgiy — CHANGES_REQUESTED** [review]

File: src/box/lua/feedback_daemon.lua
Comment (drewdzzz) [reply src/box/lua/feedback_daemon.lua:450]: I've taken a close look at this function. And I still don't see any value in guard fiber.

File: test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
Comment (drewdzzz) [reply test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua:53]: Because I wait until metrics are collected, and only then I check that they are empty (or almost empty) - the point of this test case is to check if metrics are reset after retrieving.

File: src/box/lua/feedback_daemon.lua
Comment (CuriousGeorgiy) [reply src/box/lua/feedback_daemon.lua:450]: Sorry, my bad, I missed the fact `reload` also cancels the guard loop fiber.

But anyways, AFAIC, `json.encode` can throw an error inside the `feedback_loop` (e.g., on OOM) or `send_fiber` can get accidentally canceled by the user. The latter can also happen with the `metrics_collect_loop`.

Comment (CuriousGeorgiy) [reply src/box/lua/feedback_daemon.lua:450]: Leaving it up to @alyapunov.

**Review by CuriousGeorgiy — APPROVED** [review]

File: src/box/lua/feedback_daemon.lua
Comment (alyapunov) [reply src/box/lua/feedback_daemon.lua:450]: I think one guard is enough (and perhaps even one guard is more than needed).

**Review by alyapunov — APPROVED** [review]

**Comment by drewdzzz** [comment]
@CuriousGeorgiy @alyapunov
Integration test failed because cartridge has its own [feedback wrapper](https://github.com/tarantool/cartridge/blob/master/cartridge/feedback.lua). So let's not rename fields for compatibility and file an issue in cartridge after this PR is merged - their feedback will be expired.

**Pull request merged**
